x/stakeibc/keeper/icacallbacks.go

```diff
@@ -116,3 +120,60 @@ func DelegateCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack
 	k.RecordsKeeper.RemoveDepositRecord(ctx, cast.ToUint64(recordId))
 	return nil
 }
+
+// ----------------------------------- Reinvest Callback ----------------------------------- //
+func (k Keeper) MarshalReinvestCallbackArgs(ctx sdk.Context, delegateCallback types.ReinvestCallback) ([]byte, error) {
+	out, err := proto.Marshal(&delegateCallback)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("MarshalReinvestCallbackArgs %v", err.Error()))
+		return nil, err
+	}
+	return out, nil
+}
+
+func (k Keeper) UnmarshalReinvestCallbackArgs(ctx sdk.Context, delegateCallback []byte) (*types.ReinvestCallback, error) {
+	unmarshalledReinvestCallback := types.ReinvestCallback{}
+	if err := proto.Unmarshal(delegateCallback, &unmarshalledReinvestCallback); err != nil {
+        k.Logger(ctx).Error(fmt.Sprintf("UnmarshalReinvestCallbackArgs %v", err.Error()))
+		return nil, err
+	}
+	return &unmarshalledReinvestCallback, nil
+}
+
+func ReinvestCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack *channeltypes.Acknowledgement_Result, args []byte) error {
+	// invariant: sendMsg.FromAddress == withdrawalAddress && sendMsg.ToAddress == delegationAddress
+	k.Logger(ctx).Info("ReinvestCallback executing", "packet", packet)
+
+	if ack == nil {
+		// transaction on the host chain failed
+		// don't create a record
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "ack is nil")
+	}
+
+	// deserialize the args
+	reinvestCallback, err := k.UnmarshalReinvestCallbackArgs(ctx, args)
+	if err != nil {
+		return err
+	}
+	
+	// fetch epoch
+	strideEpochTracker, found := k.GetEpochTracker(ctx, epochtypes.STRIDE_EPOCH)
+	if !found {
+		k.Logger(ctx).Error("failed to find epoch")
+		return sdkerrors.Wrapf(types.ErrInvalidLengthEpochTracker, "no number for epoch (%s)", epochtypes.STRIDE_EPOCH)
+	}
+	epochNumber := strideEpochTracker.EpochNumber
+	// create a new record so that rewards are reinvested
+	record := recordstypes.DepositRecord{
+		Id:                 0,
```

Comment:
Does the ID field not matter here cause we treat it as a list? If so, should we remove this field?

Reply:
Afaiu it gets overwritten by the latest ID when we `append` the record, so we could set it to anything here. If we can omit, we should

Reply:
ah yes, good point. Maybe makes sense to just not include an Id here (and let it default to nil) to avoid confusion

Reply:
done